Adds first cut at Distributed Closest Point query with example application

kennyweiss opened the pull request:
# Summary

- This PR is a feature
- It adds a `DistributedClosestPoint` class to quest, which solves the following problem:
    - Given an "object" mesh composed of a collection of points distributed across MPI ranks
       and a "query" mesh composed of a collection of points, distributed across the MPI ranks
    - Find the closest point in the object mesh to each point in the query mesh. (Also find the rank containing the closest point and the index of the object on that rank).
    - This can be used, e.g. to compute the distance and/or direction to the object
- The class takes the dimension (2 or 3) and the execution space (sequential CPU, openmp CPU or CUDA) as runtime parameters
- The current approach to solving the distributed problem is for each rank to generate a spatial index over its "object mesh" and to send its "query mesh" data to every other rank to update the closest point. This will be optimized in future PRs by only sending data between ranks that could affect the closest point.

- This PR also adds an example executable that uses this class. 
   - The example uses a circle as the object mesh and the vertices of a mesh in the mesh blueprint format as the query mesh
   - The latter can be generated for a Cartesian mesh in 2D or 3D or from an mfem mesh using axom's `data_collection_util`.

Here are two example runs:
```bash
## Generate a 30x30 Cartesian mesh distributed over 3 ranks 
## and run the distributed closest point query against a circular object w/ 100 samples per rank
srun -n3 ./bin/data_collection_util -p1 --min -1.5 -1.5 --max 1.5 1.5 --res 30 30
srun -n3 ./examples/quest_distributed_distance_query_ex -m box_2d.root -n 100

## Generate a Star mesh over 36 ranks, query against a circular object
cp <path/to/mfem>/data/star.mesh star.mesh
srun -n36 ./bin/data_collection_util -p1 -f 1 -m  star.mesh -l 4 --scale 1.25
srun -n36 ./examples/quest_distributed_distance_query_ex -m star.root -n 100
```

Here is a visualization of a 12 rank run on LLNL's blueos (3 nodes with 4 GPUs each) on a 300x300 Cartesian mesh:
![dcp_12rank_gpu_300x300](https://user-images.githubusercontent.com/5626552/167048795-06c4d56e-18b1-485b-b312-9a5884577dec.png)
The left side shows two derived fields of the closest point -- distance and direction to the object.
The right side shows the ranks containing the object and query points.

Here is a similar visualization of a 36 rank run on LLNL's toss3 platform on the star mesh:
![dcp_36rank_cpu_star](https://user-images.githubusercontent.com/5626552/167049293-edf8e268-0c7b-4f22-86c9-e0f5dfafc09a.png)

#### Known limitations
We will follow up with future PRs to:
* Add support for 3D queries to the application. I.e. the DistributedClosestPoint class is set up for 3D, but I haven't tested this yet
*  Optimize the MPI communications. The current code sends data from each rank to every other rank.
* Remove unnecessary memory transfers. The current code contains some extra memory copies when running on the CPU. These can be eliminated
* Improve the robustness of how the input meshes are laid out. The code currently assumes the coordinate data and computed closest_point fields have interleaved data, e.g. `xyzxyzxyz....`. It should also support coordinates in three separate stride-1 arrays, e.g. `xxx.....yyy...zzz....`

kanye-quest commented on src/axom/quest/DistributedClosestPoint.hpp:
```diff
@@ -0,0 +1,1093 @@
+  BoxArray m_boxes;
+
+#if defined(AXOM_USE_RAJA) && defined(AXOM_USE_UMPIRE)
+  SeqBVHTree* m_bvh_seq {nullptr};
```
I think `spin::BVH` should be usable without RAJA or Umpire in the sequential case.

kennyweiss replied:
Thanks -- I fixed the guards -- this query no longer requires RAJA or Umpire for the sequential policy

kanye-quest commented on src/axom/quest/DistributedClosestPoint.hpp:
```diff
@@ -0,0 +1,1093 @@
+  {
+#if defined(AXOM_USE_RAJA) && defined(AXOM_USE_UMPIRE)
+    delete m_bvh_seq;
+    m_bvh_seq = nullptr;
```
Minor: maybe consider making these BVH instance pointers `std::unique_ptr`, and deleting the destructor?

kennyweiss replied:
done

kanye-quest commented on src/axom/quest/DistributedClosestPoint.hpp:
```diff
@@ -0,0 +1,1093 @@
+
+  ~DistributedClosestPoint()
+  {
+    delete m_dcp_2;
```
Same here wrt `unique_ptr`/rule of zero

kennyweiss replied:
done

kanye-quest commented on src/axom/quest/DistributedClosestPoint.hpp:
```diff
@@ -0,0 +1,1093 @@
+ * Uses non-blocking send (isend) and blocking receives and ensures that the request
+ * objects associated with the send are finalized
+ */
+inline void send_and_recv_node(conduit::Node& send_node,
```
Have you taken a look at the `communicate_using_schema` helper class in Relay for setting up isends/irecvs?

kennyweiss replied:
Thanks @kanye-quest -- that looks promising.

@gunney1, when you're improving the MPI part of this, please look into if we can use and/or adapt `communicate_using_schema` to schedule the async send/receives.

https://github.com/LLNL/conduit/blob/3edfd63095da3e7a94ea2ae4537abb5ae7b4b547/src/libs/relay/conduit_relay_mpi.hpp#L256-L266

kanye-quest reviewed (approved):
LGTM overall 👍Just a few minor comments/suggestions.

agcapps commented on src/axom/quest/examples/quest_distributed_distance_query_example.cpp:
```diff
@@ -0,0 +1,881 @@
+      MPI_Allreduce(&inVal, &maxVal, 1, MPI_DOUBLE, MPI_MAX, MPI_COMM_WORLD);
+      MPI_Allreduce(&inVal, &sumVal, 1, MPI_DOUBLE, MPI_SUM, MPI_COMM_WORLD);
+    };
+  {
```
Why do you introduce a scope here?  And if a scope is desirable here, perhaps to group variables that aren't elsewhere used, why not put `getMinMax` inside it?

kennyweiss replied:
The scope was just to make the logging block a bit cleaner. 
I'll add a comment at the top to make it a bit clearer.

The `getMinMax` is used in a different scope below, so cannot be in this one.

agcapps commented on src/axom/quest/examples/quest_distributed_distance_query_example.cpp:
```diff
@@ -0,0 +1,881 @@
+
+    T* data = hasField(fieldName)
+      ? static_cast<T*>(
+          m_fieldsGroup->getView(axom::fmt::format("{}/values/x", fieldName))
+            ->getVoidPtr())
```
I don't understand how the `x` values give a vector field.  Please document the assumptions on data storage (such as interleaved, separate, etc.)

kennyweiss replied:
Thanks! I added a comment about this.

agcapps commented on src/axom/quest/DistributedClosestPoint.hpp:
```diff
@@ -0,0 +1,1093 @@
+   */
+  void importObjectPoints(const conduit::Node& coords, int nPts)
+  {
+    // Extract pointers to the coordinate data
+    // Note: The following assumes the coordinates are contiguous with stride NDIMS
+    // TODO: Generalize to support other strides
+
+    // TODO: Add error checking for children 'x', 'y' and 'z' and striding
```
Thank you!

agcapps commented on src/axom/quest/DistributedClosestPoint.hpp:
```diff
@@ -0,0 +1,1093 @@
+ * then pass in the object mesh and build a spatial index over this mesh.
+ * Finally, compute the closest points in the object mesh to each point in a query mesh 
+ * using the \a computeClosestPoint() function.
+ */
```
Please note mesh assumptions and preconditions, including that the object and point meshes should be decomposed over the same number of domains, and that the work is decomposed along the object mesh decomposition.  No redecomposition is done.  Thank you for the chat where we discussed this.

kennyweiss replied:
Thanks. I also added a note that the class will need to support the case where a rank has an empty object mesh (i.e. no object points), but that this is not currently supported/tested in this initial version.

agcapps commented on src/axom/quest/DistributedClosestPoint.hpp:
```diff
@@ -0,0 +1,1093 @@
+    // create conduit node containing data that has to xfer between ranks
+    conduit::Node xfer_node;
+    {
+      // clang-format off
```
thank goodness for this ability

agcapps commented on src/axom/quest/DistributedClosestPoint.hpp:
```diff
@@ -0,0 +1,1093 @@
+  }
+
+  /**
+   * \brief Computes the closest point within the objects for each query point
+   * in the provided particle mesh, provided in the mesh blueprint rooted at \a meshGroup
+   *
+   * \param meshGroup The root of a mesh blueprint for the query points
+   * \param coordset The coordinate set for the query points
```
for clarity, change `meshGroup` to `mesh_node` (see line 410)

agcapps commented on src/axom/quest/DistributedClosestPoint.hpp:
```diff
@@ -0,0 +1,1093 @@
+
+      // NOTE: uses a naive algorithm to computed distributed closest points
+      // Every rank sends it data to every other rank.
+      // TODO: Devise a more efficient algorithm to only send data to ranks with closer points
```
👍

agcapps reviewed (approved):
Thank you for this, Kenny.  Looks good.  I noted a few suggestions.  I also appreciated the bullet points you summarized as "Known Limitations," especially the last point regarding input format.  Please add comment(s) explaining this.

white238 commented on src/axom/quest/examples/quest_distributed_distance_query_example.cpp:
```diff
@@ -0,0 +1,881 @@
+
+  // clang-format off
+  const std::map<std::string, RuntimePolicy> s_validPolicies{
+    #if defined(AXOM_USE_RAJA) && defined(AXOM_USE_UMPIRE)
```
Why is RAJA and Umpire guarding seq? Shouldn't that be there regardless?  What happens when all 4 of these are off? I don't see this example guarded by any of these in the CMake but I could have missed it.

kennyweiss replied:
Thanks -- I fixed the guards -- this query no longer requires RAJA or Umpire for the sequential policy

white238 commented on src/axom/quest/examples/quest_distributed_distance_query_example.cpp:
```diff
@@ -0,0 +1,881 @@
+      mesh->GetBoundingBox(mins, maxs);
+    }
+
+    if(myRank == 0)
```
todo to move to SLIC_INFO_ROOT when my pr goes in?

kennyweiss commented on src/axom/quest/examples/quest_distributed_distance_query_example.cpp:
```diff
@@ -0,0 +1,881 @@
+  for(auto i : IndexSet(nQueryPts))
+  {
+    const auto& cp = cpPositions[i];
+    (*distances)(i) = sqrt(squared_distance(qPts[i], cp));
```
This doesn't override `i`, it just uses `i` as the loop variable.
I uses a `slam::PositionSet` as a personal preference since it has more concise notation.

I changed to variable from `i` to `idx` to (hopefully) make it a bit clearer.

white238 replied:
oh its the () operator.. got it.. carry on you don't need to change the variable name. i'm just dumb

kennyweiss replied:
I agree that syntax is a bit confusing since it is a pointer to a `GridFunction`, and it uses `operator()` instead of `operator[]`

white238 replied:
This is what happens when i look at PRs past my bedtime.

rhornung67 reviewed (approved):
This looks really good @kennyweiss

kennyweiss commented on src/axom/quest/examples/quest_distributed_distance_query_example.cpp:
```diff
@@ -0,0 +1,881 @@
+  {
+    switch(m_dc.GetMesh()->Dimension())
+    {
+    case 2:
```
Thanks. I added a `SLIC_ERROR_IF` to ensure that `DIM` can only be 2 or 3.

kennyweiss commented:
Thanks for the great feedback everyone.

I've addressed all concerns other than the unnecessary RAJA guards on the sequential policy. 
Working on that now before merging...
